internal/controller/dnsrecord_controller.go

```diff
@@ -134,36 +129,83 @@ func (r *DNSRecordReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		reason = "ValidationError"
 		message = fmt.Sprintf("validation of DNSRecord failed: %v", err)
 		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse, reason, message)
-		return r.updateStatus(ctx, previous, dnsRecord, noRequeueDuration, err)
+		return r.updateStatus(ctx, previous, dnsRecord, false, err)
 	}
 
 	// Publish the record
-	requeueAfter, err := r.publishRecord(ctx, dnsRecord)
+	hadChanges, err := r.publishRecord(ctx, dnsRecord)
 	if err != nil {
 		reason = "ProviderError"
 		message = fmt.Sprintf("The DNS provider failed to ensure the record: %v", provider.SanitizeError(err))
 		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse, reason, message)
-		return r.updateStatus(ctx, previous, dnsRecord, noRequeueDuration, err)
+		return r.updateStatus(ctx, previous, dnsRecord, hadChanges, err)
 	}
-	// success
-	dnsRecord.Status.ObservedGeneration = dnsRecord.Generation
-	dnsRecord.Status.Endpoints = dnsRecord.Spec.Endpoints
 
-	if err := r.ReconcileHealthChecks(ctx, dnsRecord); err != nil {
+	if err = r.ReconcileHealthChecks(ctx, dnsRecord); err != nil {
 		return ctrl.Result{}, err
 	}
 
-	return r.updateStatus(ctx, previous, dnsRecord, requeueAfter, nil)
+	return r.updateStatus(ctx, previous, dnsRecord, hadChanges, nil)
 }
 
-func (r *DNSRecordReconciler) updateStatus(ctx context.Context, previous, current *v1alpha1.DNSRecord, requeueAfter time.Duration, specErr error) (reconcile.Result, error) {
+func (r *DNSRecordReconciler) updateStatus(ctx context.Context, previous, current *v1alpha1.DNSRecord, hadChanges bool, specErr error) (reconcile.Result, error) {
+	var requeueTime time.Duration
+	logger := log.FromContext(ctx)
+
+	// short loop. We don't publish anything so not changing status
+	if prematurely, requeueIn := recordReceivedPrematurely(current); prematurely {
+		return reconcile.Result{RequeueAfter: requeueIn}, nil
+	}
+
+	// failure
 	if specErr != nil {
-		current.Status.WriteCounter = previous.Status.WriteCounter
-		current.Status.ValidFor = previous.Status.ValidFor
-		current.Status.QueuedAt = previous.Status.QueuedAt
-		current.Status.QueuedFor = previous.Status.QueuedFor
+		var updateError error
+		if !equality.Semantic.DeepEqual(previous.Status, current.Status) {
+			if updateError = r.Status().Update(ctx, current); updateError != nil && apierrors.IsConflict(updateError) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+		}
+		return ctrl.Result{Requeue: true}, updateError
+	}
+
+	// success
+	if hadChanges {
+		// generation has not changed but there are changes.
+		// implies that they were overridden - bump write counter
+		if !generationChanged(current) {
+			current.Status.WriteCounter++
+			writeCounter.WithLabelValues(current.Name, current.Namespace).Inc()
```

Comment:
should this involve the generation as well?

Reply:
We could :thinking:  but not sure that we should. Considering that we should have a record per listener there will be a lot happening in the metrics. I'm concerned about the dimensions of this metric. Would generation be useful considering that we reset it anyway on a generation change? I reason that whatever tool will be consuming this metric (Grafana most likely) will maintain the history to see write counts of the past gens.